src/textual/widgets/text_input.py

```diff
@@ -0,0 +1,172 @@
+from __future__ import annotations
+
+from rich.console import RenderableType
+from rich.text import Text
+
+from textual import events
+from textual._types import MessageTarget
+from textual.geometry import Size
+from textual.keys import Keys
+from textual.message import Message
+from textual.reactive import Reactive
+from textual.widget import Widget
+
+
+class TextInputBase(Widget):
+    ALLOW_PROPAGATE = {"tab", "shift+tab"}
+
+    current_text = Reactive("", layout=True)
+    cursor_index = Reactive(0)
+
+    class Changed(Message, bubble=True):
+        def __init__(self, sender: MessageTarget, value: str) -> None:
+            """Message posted when the user presses the 'enter' key while
+            focused on a TextInput widget.
+
+            Args:
+                sender (MessageTarget): Sender of the message
+                value (str): The value in the TextInput
+            """
+            super().__init__(sender)
+            self.value = value
+
+    def on_key(self, event: events.Key) -> None:
+        key = event.key
+
+        if key == "\x1b":
+            return
+
+        if key not in self.ALLOW_PROPAGATE:
+            event.stop()
+
+        changed = False
+        if key == "ctrl+h" and self.cursor_index != 0:
+            new_text = (
+                self.current_text[: self.cursor_index - 1]
+                + self.current_text[self.cursor_index :]
+            )
+            self.current_text = new_text
+            self.cursor_index = max(0, self.cursor_index - 1)
+            changed = True
+        elif key == "ctrl+d" and self.cursor_index != len(self.current_text):
+            new_text = (
+                self.current_text[: self.cursor_index]
+                + self.current_text[self.cursor_index + 1 :]
+            )
+            self.current_text = new_text
+            changed = True
+        elif key == "left":
+            self.cursor_index = max(0, self.cursor_index - 1)
+        elif key == "right":
+            self.cursor_index = min(len(self.current_text), self.cursor_index + 1)
+        elif key == "home":
+            self.cursor_index = 0
+        elif key == "end":
+            self.cursor_index = len(self.current_text)
+        elif key not in Keys.values():
+            new_text = (
+                self.current_text[: self.cursor_index]
+                + key
+                + self.current_text[self.cursor_index :]
+            )
+            self.current_text = new_text
+            self.cursor_index = min(len(self.current_text), self.cursor_index + 1)
+            changed = True
+
+        if changed:
+            self.post_message_no_wait(self.Changed(self, value=self.current_text))
+
+
+class TextInput(TextInputBase, can_focus=True):
+    CSS = """
```

Comment:
CSS is very much subject to change - I won't spend too much time tweaking things unless they're causing issues for now.